File: src/java/src/mil/navy/nrl/norm/io/NormOutputStream.java

package mil.navy.nrl.norm.io;

import java.io.IOException;
import java.io.OutputStream;
import java.net.SocketException;
import java.util.LinkedList;
import java.util.List;

import mil.navy.nrl.norm.NormEvent;
import mil.navy.nrl.norm.NormInstance;
import mil.navy.nrl.norm.NormNode;
import mil.navy.nrl.norm.NormObject;
import mil.navy.nrl.norm.NormSession;
import mil.navy.nrl.norm.NormStream;
import mil.navy.nrl.norm.enums.NormEventType;
import mil.navy.nrl.norm.enums.NormFlushMode;

/**
 * @author Jason Rush
 */
public class NormOutputStream extends OutputStream {
  private NormInstance normInstance;
  private NormSession normSession;
  private NormStream normStream;

  private List<INormEventListener> normEventListeners;

  private boolean closed;
  private Object closeLock;

  private boolean bufferIsFull;

  public NormOutputStream(String address, int port) throws IOException {
    // Create the NORM instance
    normInstance = new NormInstance();

    // Create the NORM session
    normSession = normInstance.createSession(address, port,
        NormNode.NORM_NODE_ANY);

    normStream = null;

    normEventListeners = new LinkedList<INormEventListener>();

    closed = true;
    closeLock = new Object();

    bufferIsFull = false;
  }

  public synchronized void openDebugLog(String filename) throws IOException {
    normInstance.openDebugLog(filename);
  }

  public synchronized void closeDebugLog() {
    normInstance.closeDebugLog();
  }
  
  public synchronized void setDebugLevel(int level) {
    normInstance.setDebugLevel(level);
  }
  
  public synchronized void setMessageTrace(boolean messageTrace) {
    normSession.setMessageTrace(messageTrace);
  }

  public synchronized void setMulticastInterface(String multicastInterface)
      throws IOException {
    normSession.setMulticastInterface(multicastInterface);
  }

  public synchronized void setEcnSupport(boolean ecnEnable, boolean ignoreLoss)
      throws IOException {
    normSession.setEcnSupport(ecnEnable, ignoreLoss);
  }

  public synchronized void setTtl(byte ttl) throws IOException {
    normSession.setTTL(ttl);
  }

  public synchronized void setTos(byte tos) throws IOException {
    normSession.setTOS(tos);
  }

  public synchronized void setCongestionControl(boolean ccEnabled,
      boolean ccAdjustRate) {
    normSession.setCongestionControl(ccEnabled, ccAdjustRate);
  }

  public synchronized void setTxRateBounds(double minTxRate, double maxTxRate) {
    normSession.setTxRateBounds(minTxRate, maxTxRate);
  }

  public synchronized void setTxRate(double txRate) {
    normSession.setTxRate(txRate);
  }

  public synchronized void setGrttEstimate(double initialGrttEstimate) {
    normSession.setGrttEstimate(initialGrttEstimate);
  }

  public synchronized void setGroupSize(long groupSize) {
    normSession.setGroupSize(groupSize);
  }

  public synchronized void setAutoParity(short autoParity) {
    normSession.setAutoParity(autoParity);
  }

  public synchronized void setBackoffFactor(double backoffFactor) {
    normSession.setBackoffFactor(backoffFactor);
  }

  public synchronized double getTxRate() {
    return normSession.getTxRate();
  }

  public synchronized double getGrttEstimate() {
    return normSession.getGrttEstimate();
  }

  public synchronized void setAutoFlush(NormFlushMode flushMode) {
    if (normStream == null) {
      throw new IllegalStateException(
          "Can only set auto flush after the stream is open");
    }

    normStream.setAutoFlush(flushMode);
  }

  public synchronized void setPushEnable(boolean pushEnable) {
    if (normStream == null) {
      throw new IllegalStateException(
          "Can only set push enabled after the stream is open");
    }

    normStream.setPushEnable(pushEnable);
  }

  public synchronized void markEom() {
    if (normStream == null) {
      throw new IllegalStateException(
          "Can only mark EOM after the stream is open");
    }

    normStream.markEom();
  }

  /**
   * @param normEventListener The INormEventListener to add.
   */
  public void addNormEventListener(INormEventListener normEventListener) {
    synchronized (normEventListeners) {
      normEventListeners.add(normEventListener);
    }
  }

  /**
   * @param normEventListener The INormEventListener to add.
   */
  public void removeNormEventListener(INormEventListener normEventListener) {
    synchronized (normEventListeners) {
      normEventListeners.add(normEventListener);
    }
  }

  private void fireNormEventOccured(NormEvent normEvent) {
    synchronized (normEventListeners) {
      for (INormEventListener normEventListener : normEventListeners) {
        normEventListener.normEventOccurred(normEvent);
      }
    }
  }

  public void open(int sessionId, long bufferSpace, int segmentSize,
      short blockSize, short numParity, long repairWindow) throws IOException {
    synchronized (closeLock) {
      if (!isClosed()) {
        throw new SocketException("Stream is already open");
      }

      normSession.startSender(sessionId, bufferSpace, segmentSize, blockSize,
          numParity);

      // Open the stream
      normStream = normSession.streamOpen(repairWindow);

      closed = false;
    }
  }

  /**
   * @see java.io.OutputStream#close()
   */
  @Override
  public void close() throws IOException {
    synchronized (closeLock) {
      if (isClosed()) {
        return;
      }

      normStream.close(false);
      normSession.stopSender();
      normInstance.stopInstance();

      closed = true;
    }
  }

  /**
   * @return Returns the closed.
   */
  public boolean isClosed() {
    synchronized (closeLock) {
      return closed;
    }
  }

  /**
   * @see java.io.OutputStream#write(int)
   */
  @Override
  public synchronized void write(int b) throws IOException {
    if (isClosed()) {
      throw new IOException("Stream is closed");
    }

    byte buffer[] = new byte[1];
    buffer[0] = (byte)b;

    write(buffer);
  }

  /**
   * @see java.io.OutputStream#write(byte[], int, int)
   */
  @Override
  public synchronized void write(byte[] buffer, int offset, int length)
      throws IOException {
    int n;

    if (isClosed()) {
      throw new IOException("Stream is closed");
    }

    while (length > 0) {
      while (normInstance.hasNextEvent(0, 0)) {
        processEvent();
      }

      // Wait while the buffer is full
      while (bufferIsFull) {
        processEvent();
      }

      // Write some data
      if ((n = normStream.write(buffer, offset, length)) < 0) {
        throw new IOException("Failed to write to stream");
      }

      bufferIsFull = (n == 0);

      length -= n;
      offset += n;
    }
  }

  private void processEvent() throws IOException {
    // Retrieve the next event
    NormEvent normEvent = normInstance.getNextEvent();

    // Check if the stream was closed
    if (isClosed()) {
      throw new IOException("Stream closed");
    }

    if (normEvent != null) {
      // Process the event
      NormEventType eventType = normEvent.getType();
      switch (eventType) {
        case NORM_TX_QUEUE_VACANCY:
        case NORM_TX_QUEUE_EMPTY:
          NormObject object = normEvent.getObject();
          if (!object.equals(normStream)) {
            break;
          }

          // Signal that the buffer is not full
          bufferIsFull = false;
          break;

        case NORM_TX_OBJECT_SENT:
        case NORM_TX_OBJECT_PURGED:
          normStream = null;
          break;

        default:
          break;
      }

      // Notify listeners of the norm event
      fireNormEventOccured(normEvent);
    }
  }
}

File: src/dotnet/src/Mil.Navy.Nrl.Norm/IO/NormOutputStream.cs
using System.Runtime.CompilerServices;

namespace Mil.Navy.Nrl.Norm.IO
{
    public class NormOutputStream : Stream
    {
        private NormInstance _normInstance;
        private NormSession _normSession;
        private NormStream? _normStream;

        private List<INormEventListener> _normEventListeners;

        private bool _closed;
        private object _closeLock;

        private bool _bufferIsFull;

        /// <exception cref="IOException"></exception>
        public NormOutputStream(string address, int port)
        {
            // Create the NORM instance
            _normInstance = new NormInstance();

            // Create the NORM session
            _normSession = _normInstance.CreateSession(address, port, NormNode.NORM_NODE_ANY);

            _normStream = null;

            _normEventListeners = new List<INormEventListener>();

            _closed = true;
            _closeLock = new object();

            _bufferIsFull = false;
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void OpenDebugLog(string filename)
        {
            _normInstance.OpenDebugLog(filename);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void CloseDebugLog()
        {
            _normInstance.CloseDebugLog();
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetDebugLevel(int level)
        {
            _normInstance.DebugLevel = level;
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetMessageTrace(bool messageTrace)
        {
            _normSession.SetMessageTrace(messageTrace);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetMulticastInterface(string multicastInterface)
        {
            _normSession.SetMulticastInterface(multicastInterface);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetEcnSupport(bool ecnEnable, bool ignoreLoss)
        {
            _normSession.SetEcnSupport(ecnEnable, ignoreLoss);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetTtl(byte ttl)
        {
            _normSession.SetTTL(ttl);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetTos(byte tos)
        {
            _normSession.SetTOS(tos);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetCongestionControl(bool ccEnabled, bool ccAdjustRate)
        {
            _normSession.SetCongestionControl(ccEnabled, ccAdjustRate);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetTxRateBounds(double minTxRate, double maxTxRate)
        {
            _normSession.SetTxRateBounds(maxTxRate, minTxRate);
        }

        public double TxRate
        {
            get
            {
                lock (this)
                {
                    return _normSession.TxRate;
                }
            }
            set
            {
                lock (this)
                {
                    _normSession.TxRate = value;
                }
            }
        }

        public double GrttEstimate
        {
            get
            {
                lock (this)
                {
                    return _normSession.GrttEstimate;
                }
            }
            set
            {
                lock (this)
                {
                    _normSession.GrttEstimate = value;
                }
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetGroupSize(long groupSize)
        {
            _normSession.SetGroupSize(groupSize);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetAutoParity(short autoParity)
        {
            _normSession.SetAutoParity(autoParity);
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetBackoffFactor(double backoffFactor)
        {
            _normSession.SetBackoffFactor(backoffFactor);
        }

        /// <exception cref="InvalidOperationException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetAutoFlush(NormFlushMode flushMode)
        {
            if (_normStream == null)
            {
                throw new InvalidOperationException("Can only set auto flush after the stream is open");
            }
            _normStream.SetAutoFlush(flushMode);
        }

        /// <exception cref="InvalidOperationException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void SetPushEnable(bool pushEnable)
        {
            if (_normStream == null)
            {
                throw new InvalidOperationException("Can only set push enabled after the stream is open");
            }
            _normStream.SetPushEnable(pushEnable);
        }

        /// <exception cref="InvalidOperationException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void MarkEom()
        {
            if (_normStream == null)
            {
                throw new InvalidOperationException("Can only mark EOM after the stream is open");
            }
            _normStream.MarkEom();
        }

        /// <param name="normEventListener">The INormEventListener to add.</param>
        public void AddNormEventListener(INormEventListener normEventListener)
        {
            lock (_normEventListeners)
            {
                _normEventListeners.Add(normEventListener);
            }
        }

        /// <param name="normEventListener">The INormEventListener to remove.</param>
        public void RemoveNormEventListener(INormEventListener normEventListener)
        {
            lock (_normEventListeners)
            {
                _normEventListeners.Remove(normEventListener);
            }
        }

        private void FireNormEventOccured(NormEvent normEvent)
        {
            lock (_normEventListeners)
            {
                foreach (var normEventListener in _normEventListeners)
                {
                    normEventListener.NormEventOccurred(normEvent);
                }
            }
        }

        /// <exception cref="IOException"></exception>
        public void Open(int sessionId, long bufferSpace, int segmentSize, short blockSize, short numParity, long repairWindow)
        {
            lock (_closeLock)
            {
                if (IsClosed)
                {
                    throw new IOException("Stream is already open");
                }

                _normSession.StartSender(sessionId, bufferSpace, segmentSize, blockSize, numParity);

                // Open the stream
                _normStream = _normSession.StreamOpen(repairWindow);

                _closed = false;
            }
        }

        /// <exception cref="IOException"></exception>
        public override void Close()
        {
            lock (_closeLock)
            {
                if (IsClosed)
                {
                    return;
                }

                _normStream?.Close(false);
                _normSession.StopSender();
                _normInstance.StopInstance();

                _closed = true;
            }
        }

        public bool IsClosed
        {
            get
            {
                lock (_closeLock)
                {
                    return _closed;
                }
            }
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void Write(int b)
        {
            if (IsClosed)
            {
                throw new IOException("Stream is closed");
            }

            var buffer = new byte[1];
            buffer[0] = (byte)b;

            Write(buffer);
        }

        /// <exception cref="IOException"></exception>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public override void Write(byte[] buffer, int offset, int count)
        {
            int n;

            if (IsClosed)
            {
                throw new IOException("Stream is closed");
            }

            while (count > 0)
            {
                while (_normInstance.HasNextEvent(0, 0))
                {
                    ProcessEvent();
                }

                // Wait while the buffer is full
                while (_bufferIsFull)
                {
                    ProcessEvent();
                }

                // Write some data
                if (_normStream == null || (n = _normStream.Write(buffer, offset, count)) < 0)
                {
                    throw new IOException("Failed to write to stream");
                }

                _bufferIsFull = n == 0;

                count -= n;
                offset += n;
            }
        }

        /// <exception cref="IOException"></exception>
        private void ProcessEvent()
        {
            // Retrieve the next event
            var normEvent = _normInstance.GetNextEvent();

            // Check if the stream was closed
            if (IsClosed)
            {
                throw new IOException("Stream closed");
            }

            if (normEvent != null)
            {
                // Process the event
                var eventType = normEvent.Type;
                switch (eventType)
                {
                    case NormEventType.NORM_TX_QUEUE_VACANCY:
                    case NormEventType.NORM_TX_QUEUE_EMPTY:
                        var normObject = normEvent.Object;
                        if (normObject == null || !normObject.Equals(_normStream))
                        {
                            break;
                        }

                        // Signal that the buffer is not full
                        _bufferIsFull = false;
                        break;

                    case NormEventType.NORM_TX_OBJECT_SENT:
                    case NormEventType.NORM_TX_OBJECT_PURGED:
                        _normStream = null;
                        break;

                    default:
                        break;
                }

                // Notify listeners of the norm event
                FireNormEventOccured(normEvent);
            }
        }

        public override void Flush()
        {
            throw new NotSupportedException();
        }

        public override int Read(byte[] buffer, int offset, int count)
        {
            throw new NotSupportedException();
        }

        public override long Seek(long offset, SeekOrigin origin)
        {
            throw new NotSupportedException();
        }

        public override void SetLength(long value)
        {
            throw new NotSupportedException();
        }

        public override bool CanRead => false;

        public override bool CanSeek => false;

        public override bool CanWrite => IsClosed;

        public override long Length => throw new NotSupportedException();

        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
    }
}